NPU GenAI Docs update (#32357)
openvinotoolkit/openvino: docs/articles_en/openvino-workflow-generative/inference-with-genai/inference-with-genai-on-npu.rst

         python3 -m venv npu-env

>          npu-env/bin/activate

Reviewer: don't we need 'source' in these commands?

Author: fixed

    generate models for Intel NPU. Please expect support for newer transformer versions in
    the future releases.

.. note::

    For systems based on Intel® Core™ Ultra Processors Series 2, more than 16GB of RAM
    may be required to process prompts longer than 1024 tokens with models exceeding 7B parameters,
    such as Llama-2-7B, Mistral-0.2-7B, and Qwen-2-7B.

LLM Inference on NPU
###############################################################################################

Export LLMs from Hugging Face
***********************************************************************************************

Optimum Intel is the primary way to export Hugging Face models for inference on NPU.
LLMs **must** be exported with the following settings:

* Symmetric weights compression: ``--sym``;
* 4-bit weight format (INT4 or NF4): ``--weight-format int4`` or ``--weight-format nf4``;
* Channel-wise or group-wise weight quantization: ``--group-size -1`` or ``--group-size 128``;
* Maximize the 4-bit weight ratio in the model: ``--ratio 1.0``.

**Group quantization** (GQ) with group size ``128`` is recommended for smaller models, e.g. up to

> 4B..5B parameters. Larger models may also work with group-quantization, but normally demonstrate

Author: let it be 4B-5B

a better performance with channel-wise quantization.

**Channel-wise quantization** (CW) usually performs best but may impact the model accuracy. OpenVINO
Neural Network Compression Framework (NNCF) provides various ways to compensate the quality loss,

> e.g. data-aware compression methods or GPTQ.

Reviewer: ```suggestion
**Channel-wise quantization** (CW) generally offers the best performance but may reduce model accuracy. OpenVINO
Neural Network Compression Framework (NNCF) provides several methods to compensate for the quality loss,
such as data-aware compression methods or GPTQ.
```

         :name: channel-wise-data-aware-quant

         optimum-cli export openvino -m meta-llama/Meta-Llama-3.1-8B-Instruct --weight-format int4 --sym --group-size -1 --ratio 1.0 --awq --scale-estimation --dataset wikitext2 Meta-Llama-3.1-8B-Instruct

   .. tab-item:: NF4-CW

      NF4 Symmetric data-free channel-wise compression:

      .. code-block:: console
         :name: channel-wise-data-free-quant-nf4

         optimum-cli export openvino -m meta-llama/Meta-Llama-3.1-8B-Instruct --weight-format nf4 --sym --group-size -1 --ratio 1.0  Meta-Llama-3.1-8B-Instruct

      .. note::

         Usually, NF4-CW provides a better accuracy compared to INT4-CW even with data-free compression, and data-aware methods can are also available and can further improve the compressed model accuracy.

   .. tab-item:: INT4-GQ

      INT4 Symmetric data-free group quantization:

      .. code-block:: console
         :name: group-quant

         optimum-cli export openvino -m microsoft/Phi-3.5-mini-instruct --weight-format int4 --sym --ratio 1.0 --group-size 128 Phi-3.5-mini-instruct

.. note::

   The NF4 data type is only supported on Intel® Core Ultra Processors Series 2 NPUs (formerly

>    codenamed Lunar Lake) and beyond. Please make sure to use channel-wise quantization with NF4.

Reviewer: ```suggestion
   codenamed Lunar Lake) and beyond. Use channel-wise quantization with NF4.
```

.. important::

   For the channel-wise quantization, the group size argument must be ``-1`` ("minus one"), not ``1``.

> There are pre-compressed models on Hugging Face that can be exported as-is, e.g.

Reviewer: ```suggestion
There are pre-compressed models on Hugging Face that can be exported as-is, for example:
```

> * 4-bit (INT4) `GPTQ models <https://huggingface.co/models?other=gptq,4-bit&sort=trending>`__

Reviewer: ```suggestion
* 4-bit (INT4) `GPTQ models <https://huggingface.co/models?other=gptq,4-bit&sort=trending>`__,
```

Additional configuration options
***********************************************************************************************

.. note::

   Options described in this article are NPU device-specific. Please don't expect it to work

>    with other devices!

Reviewer: ```suggestion
.. important::

   The options described in this article are specific to the NPU device and 
   may not work with other devices.
```

while potentially introducing certain usage limitations. By default, the LLM pipeline supports
input prompts up to 1024 tokens in length. It also ensures that the generated response contains
at least 128 tokens, unless the generation encounters the end-of-sequence (EOS) token or the
user explicitly sets a lower length limit for the response.

You may configure both the 'maximum input prompt length' and 'minimum response length' using
the following parameters:

* ``MAX_PROMPT_LEN`` -- defines the maximum number of tokens that the LLM pipeline can process
  for the input prompt (default: 1024),
* ``MIN_RESPONSE_LEN`` -- defines the minimum number of tokens that the LLM pipeline can generate
  in its response (default: 128).

The maximum context size for an LLM on NPU is defined as the sum of these two numbers. By default,
if the input prompt passed to the model is shorter than ``MAX_PROMPT_LEN`` tokens, time to first
token (TTFT) will still be equal to the TTFT as if the full prompt was passed, but it leaves more
room in the context for the model to generate more tokens. E.g., if the input prompt is just 30 tokens,

> the model can generate up to :math:`1024 + 128 - 32 = 1120` tokens.

Author: > as if a full-length prompt were passed

doubtful if `were` or `was` is the right form here

> why do we say 30 tokens in the example, but subtract 32?

That's a good point, `30` should've been subtracted (auto-aligned it by habit)

> the model can generate up to :math:`1024 + 128 - 32 = 1120` tokens.

Author: Applied in the commit

         ov::AnyMap pipeline_config = { { "MAX_PROMPT_LEN",  1024 }, { "MIN_RESPONSE_LEN", 512 } };
         ov::genai::LLMPipeline pipe(model_path, "NPU", pipeline_config);

In the GenAI LLM chat scenarios, the conversation history is accumulated in the context and may require
a larger ``MAX_PROMPT_LEN`` to handle the history properly.

Performance hints
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

You can configure the NPU LLM pipeline with the ``PREFILL_HINT`` and ``GENERATE_HINT`` options
to fine-tune the performance behavior. These options impact the prompt processing (first token)
and text generation (all other tokens) behavior respectively.

``PREFILL_HINT`` -- fine-tune the prompt processing stage:

* ``DYNAMIC`` (default since 2025.3) -- enables dynamic prompt execution, unblocks longer prompts.
* ``STATIC`` -- disables dynamic prompt execution, may perform better with several precise prompt sizes.

>   Default behavior until OpenVINO 2025.3.

Reviewer: ```suggestion
* ``DYNAMIC`` (default since OpenVINO 2025.3) -- enables dynamic prompt execution, supports longer prompts.
* ``STATIC`` -- disables dynamic prompt execution, may provide better performance for specific prompt sizes.
  Default behavior before OpenVINO 2025.3.
```

         ov::AnyMap pipeline_config = { { "GENERATE_HINT",  "BEST_PERF" } };
         ov::genai::LLMPipeline pipe(model_path, "NPU", pipeline_config);

Caching and ahead-of-time compilation
***********************************************************************************************

LLM compilation for NPU happens on-the-fly so may take a substantinal time. In order to
improve the user experience, the following options are available: OpenVINO Caching and Ahead-of-time
(AoT) compilation.

OpenVINO Caching
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

By caching compiled models, you can improve the initialization time of the future pipeline
runs. To do so, specify one of the following options in ``pipeline_config`` for NPU pipeline.

CACHE_DIR
-----------------------------------------------------------------------------------------------

``CACHE_DIR`` is the default OpenVINO caching mechanism. The standard OpenVINO ``CACHE_MODE``
hint defines whether the cached blob will contain its weights (``OPTIMIZE_SIZE``, results in
faster loading for group-quantized models) or not (``OPTIMIZE_SIZE``, produces a weightless blob,

> requires the original model to be present on disk).

Reviewer: ```suggestion
``CACHE_DIR`` is the default OpenVINO caching mechanism. The  ``CACHE_MODE``
hint defines how the cached blob stores weights. ``OPTIMIZE_SPEED`` includes the weights
and allows faster loading for group-quantized models.`OPTIMIZE_SIZE`` excludes the weights, 
producing a weightless blob, and requires the original model to be present on disk.
```

         ov::AnyMap pipeline_config = { { "CACHE_DIR",  ".npucache" } };
         ov::genai::LLMPipeline pipe(model_path, "NPU", pipeline_config);

NPUW_CACHE_DIR
-----------------------------------------------------------------------------------------------

``NPUW_CACHE_DIR`` is a legacy NPU-specific weightless caching option. Since OpenVINO 2025.1,
the preferred device-neutral caching mechanism is the OpenVINO caching (``CACHE_DIR``).

'Ahead-of-time' compilation
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Specifying ``EXPORT_BLOB`` and ``BLOB_PATH`` parameters works similarly to ``CACHE_DIR`` but:

* It allows to explicitly specify where to **store** the compiled model.
* For subsequent runs, it requires the same ``BLOB_PATH`` to **import** the compiled model.

> * Blob type isa also defined by ``CACHE_MODE``.

Reviewer: ```suggestion
* Blob type is also defined by ``CACHE_MODE``.
```

               ov::genai::LLMPipeline pipe(model_path, "NPU", pipeline_config);

Whisper Inference on NPU
###############################################################################################

> OpenAI Whisper support (for

Reviewer: good and maybe let's elaborate: why would I as a user need to know that? previous OV version won't work with these models? the sentence feels a bit in the air here with no end result :)

Author: Reworked this section a bit

`whisper-tiny <https://huggingface.co/openai/whisper-tiny>`__,
`whisper-base <https://huggingface.co/openai/whisper-base>`__,
`whisper-small <https://huggingface.co/openai/whisper-small>`__, or
`whisper-large <https://huggingface.co/openai/whisper-large>`__ models) was first introduced in
OpenVINO 2024.5.

Export Whisper models from Hugging Face
***********************************************************************************************

Prior to OpenVINO 2025.1 the Whisper pipeline
only accepted stateless Whisper models, exported with ``--disable-stateful`` flag:

.. code:: console

   optimum-cli export openvino --trust-remote-code --model openai/whisper-tiny whisper-tiny --disable-stateful

Since OpenVINO 2025.1, this is no longer required. Weights can remain in FP16 or be compressed in INT8:

> .. code:: consolse

Reviewer: ```suggestion
.. code:: console
```

               ov::genai::LLMPipeline pipe(model_path, "NPU", pipeline_config);

Whisper Inference on NPU
###############################################################################################

OpenAI Whisper support (for
`whisper-tiny <https://huggingface.co/openai/whisper-tiny>`__,
`whisper-base <https://huggingface.co/openai/whisper-base>`__,
`whisper-small <https://huggingface.co/openai/whisper-small>`__, or
`whisper-large <https://huggingface.co/openai/whisper-large>`__ models) was first introduced in
OpenVINO 2024.5. There are no NPU-specific requirements when running the Whisper GenAI pipeline on NPU,
so a standard OpenVINO GenAI sample works without any limitations.

Export Whisper models from Hugging Face
***********************************************************************************************

Prior to OpenVINO 2025.1 the Whisper pipeline

> only accepted stateless Whisper models, exported with ``--disable-stateful`` flag:

Reviewer 2: Why mention this at all? Not clear to reader what disable-stateful does, and for NPU usecase latest version of NPU should recommended, right?
